cpp/oneapi/dal/table/backend/homogen_table_impl.hpp

```diff
@@ -132,118 +58,87 @@ class homogen_table_impl {
         }
     }
 
-    std::int64_t get_column_count() const {
+    std::int64_t get_column_count() const override {
         return col_count_;
     }
 
-    std::int64_t get_row_count() const {
+    std::int64_t get_row_count() const override {
         return row_count_;
     }
 
-    const table_metadata& get_metadata() const {
+    const table_metadata& get_metadata() const override {
         return meta_;
     }
 
-    const void* get_data() const {
+    const void* get_data() const override {
         return data_.get_data();
     }
 
-    array<byte_t> get_data_array() const {
-        return data_;
-    }
-
-    data_layout get_data_layout() const {
+    data_layout get_data_layout() const override {
         return layout_;
     }
 
-    template <typename Data>
-    void pull_rows(array<Data>& block, const range& rows) const {
-        pull_rows_impl(detail::default_host_policy{}, block, rows, alloc_kind::host);
+    std::int64_t get_kind() const override {
+        return 1;
     }
 
-    template <typename Data>
-    void push_rows(const array<Data>& block, const range& rows) {
-        push_rows_impl(detail::default_host_policy{}, block, rows);
+    array<byte_t> get_data_array() const {
+        return data_;
     }
 
-    template <typename Data>
-    void pull_column(array<Data>& block, std::int64_t column_index, const range& rows) const {
-        pull_column_impl(detail::default_host_policy{},
-                         block,
-                         column_index,
-                         rows,
-                         alloc_kind::host);
+    template <typename T>
+    void pull_rows(const detail::default_host_policy& policy,
+                   array<T>& block,
+                   const range& rows) const {
+        homogen_pull_rows(policy, get_info(), data_, block, rows, alloc_kind::host);
     }
 
-    template <typename Data>
-    void push_column(const array<Data>& block, std::int64_t column_index, const range& rows) {
-        push_column_impl(detail::default_host_policy{}, block, column_index, rows);
+    template <typename T>
+    void pull_column(const detail::default_host_policy& policy,
+                     array<T>& block,
+                     std::int64_t column_index,
+                     const range& rows) const {
+        homogen_pull_column(policy, get_info(), data_, block, column_index, rows, alloc_kind::host);
```

Comment:
What the reason was to extract implementations of pull/push as free functions?

Reply:
To separate homogen table and homogen table builder implementations. Builder was using the `push` methods implementations from homogen table, however `push`  is not required to satisfy homogen table contract.